service/history/workflowTaskHandlerCallbacks.go

```diff
@@ -389,6 +389,14 @@ func (handler *workflowTaskHandlerCallbacksImpl) handleWorkflowTaskCompleted(
 	}
 	defer func() { workflowContext.GetReleaseFn()(retError) }()
 
+	var effects effect.Buffer
+	defer func() {
+		if retError != nil {
+			effects.Cancel(ctx)
+		}
+		effects.Apply(ctx)
+	}()
```

Comment:
This seems like the safest thing but the error handling is weird between here and workflowTaskHandler so I've also added additional calls to effects.Cancel in some key places - namely, if command/message handling fails and if the MutableState write fails.

Reply:
Why? To cancel effects earlier? I think this PR comment needs to be converted into code comment.